lightning_examples/finetuning-scheduler/finetuning-scheduler.py

```diff
@@ -0,0 +1,565 @@
+        loss = torch.stack([x["loss"] for x in outputs]).mean()
+        self.log("val_loss", loss, prog_bar=True, sync_dist=True)
+        metric_dict = self.metric.compute(predictions=preds, references=labels)
+        self.log_dict(metric_dict, prog_bar=True, sync_dist=True)
+        return loss
+
+    def init_pgs(self) -> List[Dict]:
+        """Initialize the parameter groups. Used to ensure weight_decay is not applied to our specified bias
+        parameters when we initialize the optimizer.
+
+        Returns:
+            List[Dict]: A list of parameter group dictionaries.
+        """
+        pgs = [
+            {
+                "params": [
+                    p
+                    for n, p in self.model.named_parameters()
+                    if not any(nd in n for nd in self.no_decay) and p.requires_grad
+                ],
+                "weight_decay": self.optimizer_init["init_args"]["weight_decay"],
+            },
+            {
+                "params": [
+                    p
+                    for n, p in self.model.named_parameters()
+                    if any(nd in n for nd in self.no_decay) and p.requires_grad
+                ],
+                "weight_decay": 0.0,
+            },
+        ]
+        return pgs
+
+    def configure_optimizers(self):
+        # the phase 0 parameters will have been set to require gradients during setup
+        # you can initialize the optimizer with a simple requires.grad filter as is often done,
+        # but in this case we pass a list of parameter groups to ensure weight_decay is
+        # not applied to the bias parameter (for completeness, in this case it won't make much
+        # performance difference)
+        optimizer = instantiate_registered_class(args=self.init_pgs(), init=self.optimizer_init)
+        scheduler = {
+            "scheduler": instantiate_registered_class(args=optimizer, init=self.lr_scheduler_init),
+            **self.pl_lrs_cfg,
+        }
+        return [optimizer], [scheduler]
+
+    def configure_callbacks(self):
+        found_fts = [c for c in self.trainer.callbacks if isinstance(c, FinetuningScheduler)]  # type: ignore # noqa
+        if found_fts:
+            self.finetuningscheduler_callback = found_fts[0]
+        return super().configure_callbacks()
+
+
+# %%
+# let's create a finetuning schedule for our model and run an explicitly scheduled finetuning training scenario with it
+# Please see the documentation for a full description of the schedule format
+ft_schedule_yaml = """
+0:
+  params: # the parameters for each phase definition can be fully specified
+  - model.classifier.bias
+  - model.classifier.weight
+1:
+  params: # or specified via a regex
+  - model.albert.pooler.*
+2:
+  params:
+  - model.albert.encoder.*.ffn_output.*
+"""
+ft_schedule_name = "RteBoolqModule_ft_schedule_albert_base.yaml"
+with open(ft_schedule_name, "w") as f:
+    f.write(ft_schedule_yaml)
```

Comment:
Add a comment to explain why we write this to a file.

Reply:
Great thought, done!